### ISSUE TITLE
Add batch support

### ISSUE THREAD

**Issue by sionzeecz** [share/sharedb#332]
ShareDB database really needs batch support. Which controls operations if they are success and undo previous success operation if other are failed. Something like state management.

Example:
I'm going to submit 3 ops which are connected together.
- first op was successfully executed.
- second op failed
- third op is not inserted because second failed.

It means I have in database first op, my code is looking for this value and if the value is there then it can execute code in way where it takes in account second and third op.

I expect the batch command to do something like this:
- inserts first op (success?) cool, batch it.
- trying to insert second op (failed?), cancel batch to not break workflow / or batch as failed.

So the batch should insert the values to the database only if all batched results are OK.

I can imagine to easy use it inside code:
```
const connection = getConnection();
const batch = connection.createBatch();
batch.submitOp(xxx);
batch.submitOp(xxx);
batch.submitOp(xxx);
if(batch.execute()) log("everything is synchronized and fine")
else log("Whole process was cancelled to not break workflow.")
```

If it fails during executing the batch on the server, it should undo all executed operations in the batch.

**Comment by alecgibson** [comment]
I agree that there are certain cases where a bit of atomicity is missing (mostly when attempting to set sensible values at the middleware level, eg timestamps).

What's the exact problem you're trying to solve? Is there any reason you can't just submit all the ops together as one large op?

Also note that there's a (slightly hidden) feature where you can `pause()` and `resume()` a document, which is potentially what you're after?

**Comment by sionzeecz** [comment]
In my case I'm doing few changes to doc1, then few changes to multiple other OTtypes and then again changes to doc1 (which must be done after other type changes).

If changes to other ot types fails or to the last doc1, then I have to somehow undo the all steps before.

**Comment by alecgibson** [comment]
I mean, without a concrete example it's hard to talk in non-hypotheticals, but you could also potentially try leveraging the [sub-type][1] feature of `json0`, and combine everything into a single doc, where you can apply a single, atomic op.

[1]: https://github.com/ottypes/json0#subtype-operations

**Comment by sionzeecz** [comment]
I see but I can't use json0. We have custom OTType and we really need to keep all documents separated. Anyway your idea is great and totally correct. But it just doesn't suit my requirements.
